cincinnati/src/plugins/internal/channel_filter.rs

```diff
@@ -56,14 +57,17 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = get_multiple_values!(internal_io.parameters, "channel")?.clone();
+            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
```

Comment:
I suspect we could directly fix the macro to return the proper error instead of `bail!()`, and adjust its name to make it explicitly scoped to client parameters. @steveej?
Sidenote: for the single value case here, this should be effectively equal to a `Option::ok_or()`.

Reply:
I'd rather wrap the error in the plugin so that this macros could be reused elsewhere

Reply:
How about this:


```diff
diff --git a/cincinnati/src/plugins/internal/channel_filter.rs b/cincinnati/src/plugins/internal/channel_filter.rs
index 87231eb..c26dac5 100644
--- a/cincinnati/src/plugins/internal/channel_filter.rs
+++ b/cincinnati/src/plugins/internal/channel_filter.rs
@@ -57,10 +57,9 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
-                Ok(c) => c.clone(),
-                Err(_) => return Err(GraphError::MissingParams(vec!["channel".to_string()])),
-            };
+            let channel = get_multiple_values!(internal_io.parameters, "channel")
+                .map_err(|e| GraphError::MissingParams(vec![e.to_string()]))?
+                .clone();

             if !CHANNEL_VALIDATION_REGEX_RE.is_match(&channel) {
                 return Err(GraphError::InvalidParams(format!(
diff --git a/cincinnati/src/plugins/macros.rs b/cincinnati/src/plugins/macros.rs
index 2cc216c..1911a64 100644
--- a/cincinnati/src/plugins/macros.rs
+++ b/cincinnati/src/plugins/macros.rs
@@ -9,7 +9,7 @@ macro_rules! get_multiple_values {
                     if let Some(value) = $map.get($key) {
                         value
                     } else {
-                        bail!("could not find key '{}'", $key)
+                        bail!("{}", $key)
                     },
                 )
             };
@@ -25,7 +25,7 @@ macro_rules! get_multiple_values {
                             if let Some(value) = $map.get($key) {
                                 value
                             } else {
-                                bail!("could not find key '{}'", $key)
+                                bail!("{}", $key)
                             },
                         )*
                     )
```

Reply:
How about this:


```diff
diff --git a/cincinnati/src/plugins/internal/channel_filter.rs b/cincinnati/src/plugins/internal/channel_filter.rs
index 87231eb..c26dac5 100644
--- a/cincinnati/src/plugins/internal/channel_filter.rs
+++ b/cincinnati/src/plugins/internal/channel_filter.rs
@@ -57,10 +57,9 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
-                Ok(c) => c.clone(),
-                Err(_) => return Err(GraphError::MissingParams(vec!["channel".to_string()])),
-            };
+            let channel = get_multiple_values!(internal_io.parameters, "channel")
+                .map_err(|e| GraphError::MissingParams(vec![e.to_string()]))?
+                .clone();

             if !CHANNEL_VALIDATION_REGEX_RE.is_match(&channel) {
                 return Err(GraphError::InvalidParams(format!(
diff --git a/cincinnati/src/plugins/macros.rs b/cincinnati/src/plugins/macros.rs
index 2cc216c..1911a64 100644
--- a/cincinnati/src/plugins/macros.rs
+++ b/cincinnati/src/plugins/macros.rs
@@ -9,7 +9,7 @@ macro_rules! get_multiple_values {
                     if let Some(value) = $map.get($key) {
                         value
                     } else {
-                        bail!("could not find key '{}'", $key)
+                        bail!("{}", $key)
                     },
                 )
             };
@@ -25,7 +25,7 @@ macro_rules! get_multiple_values {
                             if let Some(value) = $map.get($key) {
                                 value
                             } else {
-                                bail!("could not find key '{}'", $key)
+                                bail!("{}", $key)
                             },
                         )*
                     )
```